lib/elixir/lib/dynamic_supervisor.ex

```diff
@@ -0,0 +1,859 @@
+      case type do
+        :worker -> Map.get(child, :shutdown, 5_000)
+        :supervisor -> Map.get(child, :shutdown, :infinity)
+      end
+
+    validate_child(start, restart, shutdown, type, modules)
+  end
+
+  defp validate_child({_, start, restart, shutdown, type, modules}) do
+    validate_child(start, restart, shutdown, type, modules)
+  end
+
+  defp validate_child(other) do
+    {:invalid_child_spec, other}
+  end
+
+  defp validate_child(start, restart, shutdown, type, modules) do
+    with :ok <- validate_start(start),
+         :ok <- validate_restart(restart),
+         :ok <- validate_shutdown(shutdown),
+         :ok <- validate_type(type),
+         :ok <- validate_modules(modules) do
+      {:ok, {start, restart, shutdown, type, modules}}
+    end
+  end
+
+  defp validate_start({m, f, args}) when is_atom(m) and is_atom(f) and is_list(args), do: :ok
+  defp validate_start(mfa), do: {:invalid_mfa, mfa}
+
+  defp validate_type(type) when type in [:supervisor, :worker], do: :ok
+  defp validate_type(type), do: {:invalid_child_type, type}
+
+  defp validate_restart(restart) when restart in [:permanent, :temporary, :transient], do: :ok
+  defp validate_restart(restart), do: {:invalid_restart_type, restart}
+
+  defp validate_shutdown(shutdown) when is_integer(shutdown) and shutdown > 0, do: :ok
+  defp validate_shutdown(shutdown) when shutdown in [:infinity, :brutal_kill], do: :ok
+  defp validate_shutdown(shutdown), do: {:invalid_shutdown, shutdown}
+
+  defp validate_modules(:dynamic), do: :ok
+
+  defp validate_modules(mods) do
+    if is_list(mods) and Enum.all?(mods, &is_atom/1) do
+      :ok
+    else
+      {:invalid_modules, mods}
+    end
+  end
+
+  @doc """
+  Terminates the given child identified by child id.
+
+  If successful, this function returns `:ok`. If there is no process with
+  the given PID, this function returns `{:error, :not_found}`.
+  """
+  @spec terminate_child(Supervisor.supervisor(), pid) :: :ok | {:error, :not_found}
+  def terminate_child(supervisor, pid) when is_pid(pid) do
+    call(supervisor, {:terminate_child, pid})
+  end
+
+  @doc """
+  Returns a list with information about all children.
+
+  Note that calling this function when supervising a large number
+  of children under low memory conditions can cause an out of memory
+  exception.
+
+  This function returns a list of tuples containing:
+
+    * `id` - as defined in the child specification but is always
+      set to `:undefined` for dynamic supervisors
+
+    * `child` - the pid of the corresponding child process or the
+      atom `:restarting` if the process is about to be restarted
+
+    * `type` - `:worker` or `:supervisor` as defined in the child
+      specification
+
+    * `modules` - as defined in the child specification
+
+  """
+  @spec which_children(Supervisor.supervisor()) :: [
+          {:undefined, pid | :restarting, :worker | :supervisor, :supervisor.modules()}
+        ]
+  def which_children(supervisor) do
```

Comment:
I wonder if this function should even be open. I presume it needs to be exported for code reloading, right? Regardless, we could perhaps make it `@doc false`. Personally, I can't recall a single time I wanted to invoke `which_children` of a dynamic supervisor myself. If you do want to support that, maybe consider adding a separate `children` function which could return `[{pid | restarting, child_spec}]`?